modules/calib3d/src/usac/termination.cpp

```diff
@@ -0,0 +1,79 @@
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+#include "../precomp.hpp"
+#include "../usac.hpp"
+
+namespace cv { namespace usac {
+////////////////////////////////// STANDARD TERMINATION ///////////////////////////////////////////
+class StandardTerminationCriteriaImpl : public StandardTerminationCriteria {
+private:
+    const double log_confidence;
+    const int points_size, sample_size, MAX_ITERATIONS, MAX_TIME_MCS;
+    int predicted_iterations;
+    const bool is_time_limit;
+    std::chrono::steady_clock::time_point begin_time;
+public:
+    StandardTerminationCriteriaImpl (double confidence, int points_size_,
+                                     int sample_size_, int max_iterations_, bool is_time_limit_,
+                                     int max_time_mcs_) :
+            log_confidence(log(1 - confidence)), points_size (points_size_),
+            sample_size (sample_size_), MAX_ITERATIONS(max_iterations_),
+            MAX_TIME_MCS(max_time_mcs_), is_time_limit(is_time_limit_) {
+        predicted_iterations = max_iterations_;
+    }
+
+    /*
+     * Get upper bound iterations for any sample number
+     * n is points size, w is inlier ratio, p is desired probability, k is expceted number of iterations.
+     * 1 - p = (1 - w^n)^k,
+     * k = log_(1-w^n) (1-p)
+     * k = ln (1-p) / ln (1-w^n)
+     *
+     * w^n is probability that all N points are inliers.
+     * (1 - w^n) is probability that at least one point of N is outlier.
+     * 1 - p = (1-w^n)^k is probability that in K steps of getting at least one outlier is 1% (5%).
+     */
+    void update (const Mat &/*model*/, int inlier_number) override {
+        const double predicted_iters = log_confidence / log(1 - std::pow
+            (static_cast<double>(inlier_number) / points_size, sample_size));
+
+        // if inlier_prob == 1 then log(0) = -inf, predicted_iters == -0
+        // if inlier_prob == 0 then log(1) = 0   , predicted_iters == (+-) inf
+
+        if (! std::isinf(predicted_iters) && predicted_iters < MAX_ITERATIONS)
+            predicted_iterations = static_cast<int>(predicted_iters);
+    }
+
+    /*
+     * important to use inline here, because function is callable from RANSAC while()
+     * loop. keep function virtual, if other termination criteria do not depend on
+     * iteration or time number.
+     */
+    inline bool terminate (int current_iteration) const override {
+        // check current iteration number is higher than maximum predicted iterations.
+        if (current_iteration >= predicted_iterations) return true;
+
+        if (is_time_limit)
+            // check running time. By default max time set to maximum possible of int number.
+            return std::chrono::duration_cast<std::chrono::microseconds>
+                           (std::chrono::steady_clock::now() - begin_time).count() > MAX_TIME_MCS;
```

Comment:
IMHO, we should not introduce such time-based limits as part of OpenCV.
- no way to provide reliable testing of algorithms with such criteria (running tests in parallel on heavy load machine, mobile phones or embedded devices can get different results, etc)
- use number of iterations instead
- provide to API to configure this in User application.
- time measuring/querying on each iteration can introduce unnecessary overhead

Reply:
agree. @ivashmak, in some cases it's difficult to control it, e.g. when the computations are done in a cloud. Please, use the maximum number of iterations as hard limit.

Reply:
fixed. Now, RANSAC terminates when maximum number iterations exceed